src/poetry/console/commands/add.py

```diff
@@ -285,3 +293,22 @@ def notify_about_existing_packages(self, existing_packages: list[str]) -> None:
         for name in existing_packages:
             self.line(f"  • <c1>{name}</c1>")
         self.line(self._hint_update_packages)
+
+    def migrate_dev_dependencies(self, poetry_content: dict[str, Any]) -> None:
+        from poetry.factory import Factory
+
+        dev_group = poetry_content["group"]["dev"]["dependencies"]
+        dev_dependencies = poetry_content["dev-dependencies"]
+        for name, constraint in dev_dependencies.items():
+            canonicalized_name = canonicalize_name(name)
```

Comment:
I don't see any reason to canonicalize names as part of this migration.  

- At best it's a completely separate concern than where the dependencies are declared
- but anyway non-canonical names are fine here

Reply:
Also I think this mechanism is going to lose any comments that people have left in their dependency listing.

The code should surely move the dependencies wholesale from one place to the other, rather than reconstructing them one at a time - then I expect it should be possible to preserve comments.

Reply:
Agreed, and what I was trying to get at in https://github.com/python-poetry/poetry/pull/6677#discussion_r985172404. I still don't see why this isn't possible, but I haven't had time to pull this down locally and play with it myself.

Reply:
Non-canonical names are definitely an issue. Older versions of poetry did not handle this properly. For instance, `porg add sphinx` would add `Sphinx = "^5.3.2"` to `pyproject.toml`. If I don't use canonical names, then there is a possibility of duplicate entries. I observed this in one of my new tests:

```toml
[tool.poety.group.dev.dependencies]
Cachy = "^0.2.0"
cachy = "^0.2.0"
```

Not good ☹️

Reply:
if I have a `pyproject.toml` that asks for `Cachy` then I will be surprised and perhaps disappointed if this migration changes that to `cachy` - it simply shouldn't need to do that.

I don't understand where your two `cachy`s come from.  Are you trying to `poetry add cachy` to a project that already holds `Cachy`?  or what?

Conceptually this MR should be simpler than you are making it I think:

- if `poetry.group.dev.dependencies` and `poetry.dev-dependencies` are both present then bail, perhaps with a log saying that the user probably wants to sort that out manually
- it `poetry.dev-dependencies` exists, then pop that out of the dictionary and re-insert it as `poetry-group.dev.dependencies`
- that's it

Reply:
Non-canonical package names come from older versions of poetry. For example, `sphinx` is added as `Sphinx`:
```toml
[tool.poetry.dev-dependencies]
pytest = "^5.2"
Sphinx = "^5.2.3"
```

Reply:
yes, but so what?  that doesn't answer either (i) why you are getting two names for `cachy` or (ii) why this migration should canonicalize those names

Reply:
I think I see the confusion in your unit tests.

After going `app.poetry.file.write(content)` you also need to `app.reset_poetry()` or else explicitly update the `poetry` object that the tests are passing around with the new dependency.

Once you do that you'll find that you don't need to do the same from within poetry src.

Then indeed the entire implementation of `migrate_dev_dependencies` can be
```python
        dev_dependencies = poetry_content.pop("dev-dependencies")
        poetry_content["group"]["dev"]["dependencies"] = dev_dependencies
```
(which is hardly worth a separate method IMO)

Reply:
@dimbleby As I said earlier, the `dev` group doesn't get created when there is the `dev-dependencies` section. The rest of the `add` command seems to need that to function properly. Therefore, I think I still have to have the loop. As for why the two `cachy`'s, I have seen something similar when migrating from 1.1x to 1.2. I'm trying to come up with a better example to illustrate this.

Thanks for the tip on `app.reset_poetry()`. I'll try that.

Reply:
Here's what I'm seeing with the double `cachy`:
```python
tests/console/commands/test_add.py:2213: AssertionError


>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>> traceback >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>

name = 'Cachy'
app = <tests.helpers.PoetryTestApplication object at 0x7f30401fcf40>
repo = <tests.helpers.TestRepository object at 0x7f303f833130>
tester = <cleo.testers.command_tester.CommandTester object at 0x7f303f809ab0>

    @pytest.mark.parametrize("name", ["cachy", "Cachy"])
    def test_add_existing_constraint_old_dev_section_migrates(
        name: str, app: PoetryTestApplication, repo: TestRepository, tester: CommandTester
    ):
        # Note: Not testing non-canonical names since this is currently broken
        # for this scenario
    
        repo.add_package(get_package("cachy", "0.1.0"))
        repo.add_package(get_package("cachy", "0.2.0"))
    
        content = app.poetry.file.read()
        content["tool"]["poetry"]["dev-dependencies"] = {name: "^0.2.0"}
        app.poetry.file.write(content)
        app.reset_poetry()
    
        tester.execute("cachy^0.2.0 -G dev")
    
        warning = """\
    The dev-dependencies section is deprecated. \
    Automatically rewriting dev-dependencies as \
    group.dev.dependencies
    """
        expected = """\
    
    Updating dependencies
    Resolving dependencies...
    
    Writing lock file
    
    Package operations: 1 install, 0 updates, 0 removals
    
      • Installing cachy (0.2.0)
    """
    
        assert tester.io.fetch_output() == expected
        assert tester.io.fetch_error() == warning
    
        content = app.poetry.file.read()["tool"]["poetry"]
    
        assert "dev-dependencies" not in content
        assert "cachy" in content["group"]["dev"]["dependencies"]
>       assert content["group"]["dev"]["dependencies"] == {"cachy": "^0.2.0"}
E       AssertionError: assert {'Cachy': '^0.2.0', 'cachy': '^0.2.0'} == {'cachy': '^0.2.0'}
E         Common items:
E         {'cachy': '^0.2.0'}
E         Left contains 1 more item:
E         {'Cachy': '^0.2.0'}
E         Full diff:
E         - {'cachy': '^0.2.0'}
E         + {'Cachy': '^0.2.0', 'cachy': '^0.2.0'}
```

The call to `app.reset_poetry()` didn't help for the test that where `Cachy = "^0.2.0"` is in `[tool.poetry.dev-dependencies]`. Also, I see other writes to `pyproject.toml` in this file, and none of them are calling that method. For example, [this](https://github.com/python-poetry/poetry/blob/master/tests/console/commands/test_add.py#L1000).